### PR TITLE
Add `substitute_node` method to `DAGCircuit`

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
An attempt at implementing `substitute_node` in Rust.

### CODE REVIEW

File: crates/circuit/src/dag_circuit.rs
Comment (ElePT) [inline crates/circuit/src/dag_circuit.rs:2446]: I am not sure how to get a mutable reference to `packed` so that we can replace the node info, as it comes from `node_weight` which is returned immutably from `self.dag.node_weight(node_id)`

File: crates/circuit/src/dag_circuit.rs
Comment (ElePT) [inline crates/circuit/src/dag_circuit.rs]: not sure why these signatures got formatted... but I can revert them once the PR is ready to merge.

File: crates/circuit/src/dag_circuit.rs
Comment (ElePT) [inline crates/circuit/src/dag_circuit.rs:2482]: It seems to be not possible to clone `new_op`, and it is being moved into `new_weight`, so I am re-generating it here for lack of a better solution.

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:2695]: If we can, we should avoid cloning the node here.

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:2704]: It looks like we're missing the check to ensure that `old_packed.op.num_{qubits,clbits}()` is the same as `new_op.op.num_{qubits,clbits}()`.

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:2709]: I think it'd be best to only use `node` to get the ID of the node the user is referring to, and otherwise use `old_packed` to read any information. This way, the data in `self.dag` is the single source of truth (i.e. this way, the user can't get a weird outcome by crafting a `node` with an ID and payload that don't match what is in our graph).

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:2721]: You should be able to just do `old_packed.condition()`, as of the latest changes in `oxidize-dag`.

**Review by kevinhartman — APPROVED** [review]
Thanks @ElePT, this looks like a good start.

I think we will still need some of the ugly manual condition / wire checking from the original Python implementation.

I'm going to merge this as-is in the interest of time, since it's one of the bits needed to unblock testing for the joint PR branch. Matt or myself can add any missing checks / condition stuff to the final PR.

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(name = "nodes_on_wire", signature = (wire, only_ops=false))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature=(op=None, include_directives=true))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (remove_idle_qubits=false))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (node, op, inplace=false, propagate_condition=true))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (node, input_dag, wires=None, propagate_condition=true))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (node_block, op, wire_pos_map, cycle_check=true))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (other, qubits=None, clbits=None, front=false, inplace=true))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (*qregs))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (*qubits))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (*cregs))]
```

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs]:
```suggestion
    #[pyo3(signature = (*clbits))]
```

**Review by kevinhartman — COMMENTED** [review]
Reverting formatting changes to avoid causing any superfluous merge issues for others working on the shared branch.

**Pull request merged**
